### PR TITLE
Verify agreement messages concurrently

### LINKED ISSUE #14: Optimize agreement messages verification
**Describe the bug**
As per the golang impl, consensus (Accumulator) should be capable of verifying agreement messages concurrently or in parallel. By the default, golang impl spawns 4 workers. This could be achieved by implementing any of the following approaches:

1  Run a separate `tokio::runtime` with 4 dedicated threads to perform solely agreement message verification.
2. Use `tokio::spawn_blocking()` or native thread::spawn()
3. Use data-parallelism library (e.g `Rayon`)

NB. For options 2 and 3, we'll need to bridge with sync code.

**To Reproduce**
Running consensus with 64 provisioners (respectively more agreement messages exchanged) would create a perf bottleneck in the process of agreements verification .
 

**Additional context**
With current impl, there is a noticeable consensus perf downgrade when running `testbed` with  8 or more provisioners.

### PR DESCRIPTION
**Pull request by goshawk-3**
fixes #14 #18 

Test params:
- Single-process testbed of 10 consensus instances.
- 10 active provisioners
-  Executed for 1000 rounds
- Config:  CONSENSUS_DELAY_MS  = 1s;

Results

```
INFO testbed: bls_key=b1a78b0a33bd55f4, round=999,   average_block_time=2.50
INFO testbed: bls_key=80d3ba2aa06cc70f, round=999,   average_block_time=2.50
INFO testbed: bls_key=8bc4ce78b3549a1b, round=999,   average_block_time=2.50
INFO testbed: bls_key=89087d7526fc9515, round=999,   average_block_time=2.50
INFO testbed: bls_key=94c3bb0dc9eb7dbb, round=999,   average_block_time=2.50
INFO testbed: bls_key=b6161ab9b21bdbd3, round=999,   average_block_time=2.50
INFO testbed: bls_key=8c9eccf33d595953, round=999,   average_block_time=2.50
INFO testbed: bls_key=8a123d6b269a8458, round=999,   average_block_time=2.50
INFO testbed: bls_key=a773bb2a84a71c22, round=999,   average_block_time=2.50
INFO testbed: bls_key=928f43a5d976b469, round=999,   average_block_time=2.50
```

### CODE REVIEW

File: src/util/pubkey.rs [hunk 1]
```diff
@@ -13,56 +13,66 @@ use std::cmp::Ordering;
...
+    }
+}
+
 impl PartialOrd<PublicKey> for PublicKey {
     fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        self.0
-            .pk_t()
-            .to_bytes()
-            .partial_cmp(&other.0.pk_t().to_bytes())
+        self.to_bytes().partial_cmp(&other.1)
```
Comment (herr-seppia) [inline src/util/pubkey.rs:61]: To keep consistency, consider to have both side equals
So use always `.1` or `to_bytes()`

File: src/util/pubkey.rs [hunk 2]
```diff
@@ -13,56 +13,66 @@ use std::cmp::Ordering;
...
     }
 }
 
 impl Ord for PublicKey {
     fn cmp(&self, other: &Self) -> Ordering {
-        self.0.pk_t().to_bytes().cmp(&other.0.pk_t().to_bytes())
+        self.to_bytes().cmp(&other.1)
```
Comment (herr-seppia) [inline src/util/pubkey.rs:67]: Ditto

**Review by herr-seppia — CHANGES_REQUESTED** [review]
LGTM overall,
but I need the `Drop` comment to be address first

File: src/agreement/accumulator.rs [hunk 3]
```diff
@@ -28,63 +28,104 @@ type StorePerHash = HashMap<Hash, AgreementsPerStep>;
...
 
 impl Accumulator {
-    pub fn new(
-        _workers_amount: usize,
-        collected_votes_tx: Sender<Message>,
-        committees_set: Arc<Mutex<CommitteeSet>>,
-        ru: RoundUpdate,
-    ) -> Self {
-        let (tx, mut rx) = mpsc::channel::<Message>(100);
+    pub fn new() -> Self {
```
[the remark this answers is not in the document]
Comment (goshawk-3) [reply src/agreement/accumulator.rs]: Good point. However this function should accept queue cap as input.  I updated it accordingly.

File: src/agreement/verifiers.rs [hunk 4]
```diff
@@ -121,13 +117,13 @@ async unsafe fn aggregate_pks(
         let guard = committees_set.lock().await;
         let provisioners = guard.get_provisioners();
 
-        for m in subcomittee.into_iter() {
-            if let Some(m) = provisioners.get_member(&m.0) {
+        for (pubkey, _) in subcomittee.into_iter() {
+            if let Some(m) = provisioners.get_member(&pubkey) {
```
[the remark this answers is not in the document]
Comment (goshawk-3) [reply src/agreement/verifiers.rs]: Done

File: src/consensus.rs [hunk 5]
```diff
@@ -124,65 +192,16 @@ impl Consensus {
...
-
-        // Tear-down procedure
-        // TODO: Delete all candidates related to this round execution
-
-        // Cancel all tasks
-        agreement_task_handle.abort();
-        main_task_handle.abort();
-
-        result
+    async fn send_agreement(mut agr_inbound_queue: PendingQueue, msg: Message) {
```
[the remark this answers is not in the document]
Comment (goshawk-3) [reply src/consensus.rs]: Done

File: src/util/pubkey.rs (hunk 1 above)
Comment (goshawk-3) [reply src/util/pubkey.rs]: Done

File: src/util/pubkey.rs (hunk 2 above)
Comment (goshawk-3) [reply src/util/pubkey.rs]: Done

File: src/agreement/accumulator.rs [hunk 6]
```diff
@@ -28,63 +28,104 @@ type StorePerHash = HashMap<Hash, AgreementsPerStep>;
 
 pub(crate) struct Accumulator {
     workers: Vec<JoinHandle<()>>,
-    inbound: Sender<Message>,
+    tx: async_channel::Sender<Message>,
+    rx: async_channel::Receiver<Message>,
```
[the remark this answers is not in the document]
Comment (goshawk-3) [reply src/agreement/accumulator.rs:32]: Not sure if this makes sense. As we abort tasks, the atomic reference counter is zero-ed and these are freed on their drop too.

File: src/consensus.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (herr-seppia) [reply src/consensus.rs]: My intention was to have a reference for the message too, to avoid `clone` twice in a row

**Review by herr-seppia — APPROVED** [review]
LGTM
We'll take care of refs optimizations in a following PR

**Pull request merged**